Conditionally resize LHS Tensors before RHS TensorExpression evaluation

macedo22 opened the pull request:
fixes #4155 

## Proposed changes

In the case where the data type of the `Tensor`s are `DataVector` and `TensorExpression` splitting is used to evaluate subexpressions of the RHS, this PR makes it so that LHS `Tensor`s are sized before starting RHS `TensorExpression` evaluation instead of depending on an assignment with `=` from the result of a RHS subtree.

In almost every case, the result of a subtree will be a `DataVector` expression that can be directly assigned to the LHS component, which is why the current method of assigning to a result from a subexpression has worked. However, the example in #4155 is a case where the LHS `Tensor` components were being assigned to the result of the subexpression containing only a `NumberAsExpression` leaf - it was using the numerator `1.0` to initialize the LHS components, which can't properly size the LHS components with `=` assignment. This case could be specifically handled, but I chose the safer route of just sizing the LHS components at the outset (if necessary) and not depending on a RHS subexpression result to avoid running into another edge case like #4155 that I hadn't considered.
 
### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

macedo22 committed:
Make TensorExpressions store height relative to nearest Tensor

macedo22 committed:
Resize LHS Tensors before RHS TensorExpression evaluation if needed

nilsdeppe commented on src/DataStructures/Tensor/Expressions/AddSubtract.hpp:
```diff
@@ -463,6 +463,20 @@ struct AddSub<T1, T2, ArgsList1<Args1...>, ArgsList2<Args2...>, Sign>
     }
   }
 
+  /// \brief Get a component from a `Tensor` in this expression's subtree of the
```
I don't think this documentation is correct, or I'm not understanding it at all

nilsdeppe commented on src/DataStructures/Tensor/Expressions/AddSubtract.hpp:
```diff
@@ -463,6 +463,20 @@ struct AddSub<T1, T2, ArgsList1<Args1...>, ArgsList2<Args2...>, Sign>
+  ///
+  /// \return a component from a `Tensor` in this expression's subtree of the
+  /// RHS `TensorExpression`
+  SPECTRE_ALWAYS_INLINE auto get_used_for_size() const {
```
I don't like this name. It's not very descriptive. What about `get_size_of_members` or `get_component_size`? I have a preference for the second, but any better names are also fine with me

macedo22 replied:
Ok, my only hangup with the alternative suggestions is that the function returns a component (e.g. `DataVector`), not a size (e.g. `5`), so I wouldn't want it to be called `get_component_size` when it returns a component, not the size. Do you have a preference for (1) using the name `get_component_size` and then I update the function to return the size instead of a component or (2) using a name like `get_rhs_tensor_component` and keeping the return type as-is?

nilsdeppe replied:
Oh yea, why not return a size?

macedo22 replied:
Ok, I'll do that! I assume then the return type should be `const auto` instead of `const auto&` since it will be an integral type?

nilsdeppe replied:
just do `size_t`. `Utilities/ContainerHelpers.hpp` has a `get_size` function that you can call

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -585,6 +585,15 @@ struct TensorContract
     }
   }
 
+  /// \brief Get a component from a `Tensor` in this expression's subtree of the
```
Wrong docs?

macedo22 replied:
Can you explain what you mean by your question? I'm so sorry, I'm not following

nilsdeppe replied:
I understand now, I guess the function just doesn't seem useful based on the documentation. Returning a random tensor component that you don't know what it is seems a bit silly. I think it's better to return the size.

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Divide.hpp:
```diff
@@ -179,6 +179,20 @@ struct Divide : public TensorExpression<
     }
   }
 
+  /// \brief Get a component from a `Tensor` in this expression's subtree of the
```
Wrong docs?

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -203,10 +203,29 @@ void evaluate_impl(
       "dimensions. Check RHS and LHS indices that use the same generic index.");
+  static_assert(Derived::height_relative_to_closest_tensor_leaf_in_subtree <
+                    std::numeric_limits<size_t>::max(),
+                "This either indicates that no Tensors were found in the RHS "
```
"This" is not a great word in general, since it makes reading more difficult because of the implied context. Here, I don't understand what "this" refers too. What about `Either no Tensors were found...` and then `maximum size_t value (extremely unlikely)`

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -203,10 +203,29 @@ void evaluate_impl(
+                "This either indicates that no Tensors were found in the RHS "
+                "TensorExpression or that the depth of the tree exceeded the "
+                "maximum size_t value. If there is indeed a Tensor in the RHS "
+                "expression, this indicates a flaw in the logic for the "
```
`expression, this...` -> `expression, then the logic for ...`

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -203,10 +203,29 @@ void evaluate_impl(
+                "expression, this indicates a flaw in the logic for the "
+                "derived TensorExpression types' member, "
+                "height_relative_to_closest_tensor_leaf_in_subtree.");
 
```
Is there a `static_assert` somewhere already that the type `X` must be either `DataVector` or a fundamental? If not, this code certainly makes it a requirement. It should be noted in that assertion that the `DataVector` restriction can be generalized to other vector types (we have modal and complex(modal) vectors)

macedo22 replied:
No it isn't, and I think that is something that is long overdue, because there are several things besides this change that would break if someone used a different data type. I will add this assert

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Negate.hpp:
```diff
@@ -131,6 +131,15 @@ struct Negate
     }
   }
 
+  /// \brief Get a component from a `Tensor` in this expression's subtree of the
```
Wrong docs?

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Product.hpp:
```diff
@@ -208,6 +208,20 @@ struct OuterProduct<T1, T2, IndexList1<Indices1...>, IndexList2<Indices2...>,
     }
   }
 
+  /// \brief Get a component from a `Tensor` in this expression's subtree of the
```
Wrong docs?

nilsdeppe commented on src/DataStructures/Tensor/Expressions/SquareRoot.hpp:
```diff
@@ -140,6 +140,15 @@ struct SquareRoot
     }
   }
 
+  /// \brief Get a component from a `Tensor` in this expression's subtree of the
```
Wrong docs?

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorAsExpression.hpp:
```diff
@@ -296,6 +296,11 @@ struct TensorAsExpression<Tensor<X, Symm<SymmValues...>, IndexList<Indices...>>,
     }
   }
 
+  /// \brief Get a component from the `Tensor` contained by this expression
```
Wrong docs?

nilsdeppe commented on src/DataStructures/Tensor/Expressions/AddSubtract.hpp:
```diff
@@ -463,6 +463,20 @@ struct AddSub<T1, T2, ArgsList1<Args1...>, ArgsList2<Args2...>, Sign>
+  ///
+  /// \return a component from a `Tensor` in this expression's subtree of the
+  /// RHS `TensorExpression`
+  SPECTRE_ALWAYS_INLINE auto get_used_for_size() const {
```
all return types to `const auto&`

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -166,6 +176,11 @@ struct Expression {};
 /// be used instead of `tenex::evaluate`. See the documentation for
 /// `tenex::update` for more details and `tenex::detail::evaluate_impl` for why
 /// this is safe to do.
+/// - function `auto get_used_for_size() const`: Gets a component from a
```
`const auto&`

macedo22 committed:
Restrict Tensor data type in TensorExpressions to double, DataVector

Support for other data types can be added. We restrict it to double and DataVector for now because that is what is currently supported.

macedo22 commented:
@nilsdeppe Pushed a fixup commit and a separate commit for asserting people don't use unsupported data types

wthrowe reviewed (changes requested):
Add a test.  Perhaps the expression from the bug report.

macedo22 commented:
> Add a test.  Perhaps the expression from the bug report.

Edit: To spare you from reading everything else below, the TLDR is that encountering this edge case is very implementation dependent and to guard against implementation changes that could make a test case no longer test what we want, I think maybe I can just make a mock derived `TensorExpression` type that behaves how we want and ensures the edge case behavior we want to test actually is encountered. I will try that unless you have a better idea!

----------------

Happy to add the expression from the big report! However, I purposely didn't add a new test case in this bug fix because the test would be super implementation-dependent and I don't know how to guard against implementation changes.

If later the TE implementation is updated and it changes the structure of the tree, where it's split, the control flow, or otherwise where/when/how the LHS components are first initialized, then it's possible the test case added won't be testing the edge case at all anymore. Essentially, the test would need to assert that the first assignment to the LHS components is a double, and when/where/how this first assignment happens is very implementation-dependent. For example, if the implementation for `Divide::evaluate_primary_children()` changed so that the denominator is evaluated before the numerator, we would no longer encounter the edge case we are wanting to test, and I'm not sure how to guard against that with an assert or something in the test case.

I don't think we shouldn't add a test, I just don't know what the solution is to it being dependent on such low-level TE implementation details. I'll keep thinking about this, but let me know your thoughts on this!

macedo22 commented:
@wthrowe I made a mock division tensor expression that is a snapshot of what the implementation is right now so that we can use it in the test and know it will get at the edge case behavior that caused the bug in the first place. I'm not loving my idea of wholesale copying over the current whole class for one test case though, so let me know if you have any better ideas!

nilsdeppe commented:
I personally would just test the case that @wthrowe suggested. If you want to be sure edge cases are caught, look at the code coverage. I think that'll tell you at least that all paths are tested.

macedo22 commented:
@nilsdeppe Ok I can do that. I want to make sure I note that I don't think coverage of the code path that contains the bug is enough to elicit the bug that was found, because that same path can be taken with a different expression and not cause the bug that Will found. So I think it is better than nothing, but I don't think it's sufficient for catching this bug.

wthrowe commented:
Yes, I think that's fine.  Regression tests usually have that limitation, but for this sort of thing I think they're the best reasonable solution.

Go ahead and squash.  The new fixup is the changes in the test directory, so easy enough for anyone to review separately if they want to.

macedo22 commented:
@wthrowe @nilsdeppe Sorry to dismiss your reviews, but I found a small fixup that I think belongs with this bug fix.

I saw I had documentation explaining how I had cleverly prevented pushing flawed code with the bug Will found by deleting a member function, but I pranked myself. Therefore, the comments there I had are false and I so removed them. In addition to this, I realized that the function I had commented on and deleted really actually should never be called but for an entirely different reason. So the fixup also explains the correct reason the deleted functions should be deleted - that they will never be a point in the tree where we split.

nilsdeppe commented:
LGTM!

wthrowe commented:
Squash.
